ESQL: Parallelize newline split discovery for large text files (#155009)
elastic/elasticsearch: docs/reference/query-languages/esql/esql-data-federation-datasets.md

| `error_mode` | `fail_fast` | How malformed rows are handled. Valid values: `"fail_fast"`, `"skip_row"`, `"null_field"`. For Parquet, `skip_row` fills affected columns with null instead of skipping the entire row. For CSV, TSV, and NDJSON, `null_field` fills only individual value failures with null. Rows whose structure cannot be parsed (for example, an unparsable JSON line or a malformed CSV row) are still dropped. |
| `max_errors` | unbounded | Maximum malformed rows allowed before the query fails. Ignored when `error_mode` is `fail_fast`. |
| `max_error_ratio` | `0.0` | Fraction of malformed rows allowed (0.0–1.0). Ignored when `error_mode` is `fail_fast`. |

> | `target_split_size` | `64mb` | The target size of each unit of work a file is divided into to be read in parallel across nodes. A file larger than this is cut into several splits; a smaller file is read as a single split. Lower it for more parallelism over a few large files, raise it to reduce planning work over a very large number of bytes. |

Reviewer: Minor suggestion to shorten sentences
```suggestion
| `target_split_size` | `64mb` | The target size of each unit of work a file is divided into to be read in parallel across nodes. Files larger than the target are cut into several splits. Files smaller than the target are read as a single split. Lower the split size for more parallelism over a few large files. Raise it to reduce planning work over a very large number of bytes. |
```

Author: Done

> | `split_probe_window` | `256kb` | The bytes one search for a record boundary may read while dividing files into splits. A dataset whose records are longer than this is cut into fewer splits than `target_split_size` asks for, and reads with less parallelism, because a search that does not reach the end of a record finds no boundary to split at. Raise it for a dataset with long records, within the budget in the note below. It applies to NDJSON and to CSV and TSV without quoting or escaping: quoted or escaped records cannot be searched at a fixed offset, so those files are scanned sequentially and bounded by `external_max_record_size`, as is dividing a split further across the threads of one node. Values below about `136kb` are read in full by every search, since finishing a window that small costs less than opening another connection, so lowering it that far raises the bytes read instead of lowering them. |

Reviewer: I suggested changing "longer" to "larger" since it's a byte value not a duration but if I've misunderstood LMK. I also tried to shorten sentences and clarify some occurrences of "it". I also dropped what seemed like unnecessary details at the end:
```suggestion
| `split_probe_window` | `256kb` | The number of bytes a search for a record boundary can read while dividing files into splits. A dataset with records that are larger than this value is cut into fewer splits than the `target_split_size`. In this case, reads occur with less parallelism because a search that does not reach the end of a record finds no boundary to split at. Raise the value for a dataset with long records, within the budget in the note that follows. The setting applies to NDJSON and to CSV and TSV without quoting or escaping. Quoted or escaped records cannot be searched at a fixed offset, so those files are scanned sequentially and bounded by `external_max_record_size`, as it is dividing a split further across the threads of one node. Values below about `136kb` are read in full by every search, since finishing a window that small costs less than opening another connection. |
```

Author: Done

> | `max_split_probes` | `1000` | The maximum number of record-boundary searches a query may perform, which bounds how many splits its files are cut into: a searched file yields one split more than the searches spent on it, and a file too small to search is read as a single whole-file split. A scan asking for more splits than this is read at a wider split size than `target_split_size` requests. Raise it to get the requested split size on a very large scan. The highest accepted value is `10000`. |

Reviewer: Suggest splitting into smaller phrases:
```suggestion
| `max_split_probes` | `1000` | The maximum number of record-boundary searches a query can perform, which bounds how many splits its files are cut into. A searched file yields one split more than the searches spent on it. A file too small to search is read as a single whole-file split. A scan asking for more splits than this setting value is read at a wider split size than `target_split_size` requests. Raise it to get the requested split size on a very large scan. The highest accepted value is `10000`. |
```

Author: Done

:::{note}

> `max_split_probes` and `split_probe_window` are independent: the first bounds how many record-boundary searches a query runs, the second how many bytes each one reads. Their product is the bytes a query may read while searching, and it may not exceed 4 GB; at the defaults it is 1000 searches of `256kb`, or 256 MB. Size the window from the dataset's longest record and the count from the number of splits the scan needs, and lower one of them if the pair is rejected. The budget covers searches at fixed offsets: a sequentially scanned file (quoted or escaped CSV and TSV) is bounded by `external_max_record_size` rather than by either key.

Reviewer: More minor shortening
```suggestion
`max_split_probes` and `split_probe_window` are independent. The first defines how many record-boundary searches a query runs. The second defines how many bytes each one reads. Their product is the bytes a query can read while searching, which cannot exceed 4 GB. With the default values, it is 1000 searches of `256kb`, or 256 MB. Size the window from the dataset's longest record and the count from the number of splits the scan needs. Lower one of them if the pair is rejected. The budget covers searches at fixed offsets: a sequentially scanned file (quoted or escaped CSV and TSV) is bounded by `external_max_record_size` rather than by either key.
```

Author: Done
